### PR TITLE
Commentary in admin page

### PR DESCRIPTION
**Pull request by kbats183**
#42

### CODE REVIEW

File: backend/src/main/kotlin/org/icpclive/cds/clics/ClicsDataSource.kt [hunk 1]
```diff
@@ -125,26 +130,48 @@ class ClicsDataSource : ContestDataSource {
                             }
                         }
                     }
+                    is CommentaryEvent -> {
+                        analyticsEventsFlow.emit(
+                            AnalyticsCommentaryEvent(
+                                it.data.id,
+                                it.data.message,
+                                it.data.contest_time,
+                                it.data.team_ids?.firstOrNull()
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/clics/ClicsDataSource.kt:139]: Maybe we should put all the list in API object?

File: backend/src/main/kotlin/org/icpclive/cds/clics/ClicsDataSource.kt [hunk 2]
```diff
@@ -125,26 +130,48 @@ class ClicsDataSource : ContestDataSource {
...
+            launchICPCServices(rawRunsFlow, contestInfoFlow, analyticsEventFlow)
         }
     }
 
     override suspend fun loadOnce() = coroutineScope {
-        val (_, contestInfoFlow) = launchLoader()
+        val (_, contestInfoFlow, analyticsEventFlow) = launchLoader()
+        val analyticsEvents = merge(contestInfoFlow, analyticsEventFlow)
+            .takeWhile { it !is ContestInfo || it.status != ContestStatus.OVER }
+            .fold(mutableListOf<AnalyticsEvent>()) { ac, it ->
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/clics/ClicsDataSource.kt:162]: .filterIsInstance<AnalyticsEvent>
.mapNotNull { it as? AnalyticsEvent }

Not sure which of them is defined on flows. On the other side, I don't understand what you are trying to do. Thay can be merged in any way, not same as was in original flow, so filtering merge until some condition on one of them doesn't make much sense.

File: backend/src/main/kotlin/org/icpclive/admin/Api.kt [hunk 3]
```diff
@@ -83,9 +82,12 @@ internal inline fun <reified SettingsType : ObjectSettings, reified WidgetType :
         }
         DataBus.adminActionsFlow.emit(call.request.uri)
     }
-    post("/show_with_ttl") {
+    post("/create_and_show_with_ttl") {
         call.adminApiAction {
-            presets.showWithTTL(call.safeReceive(), call.request.queryParameters["ttl"]?.toLong())
+            return@adminApiAction presets.createAndShowWithTtl(
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/admin/Api.kt:87]: Why this return changes anything?

File: admin/src/components/AnalyticsEvent.js [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    const tickerService = useMemo(() => new PresetWidgetService("/tickerMessage", console.log), []);
+    const [advertisementTtl, setAdvertisementTtl] = useState(30);
+    const [tickerMsgTtl, setTickerMsgTtl] = useState(120);
+
+    useEffect(() => {
+        ws.current = new WebSocket(apiUrl());
+
+        ws.current.onmessage = (event) => {
+            const data = JSON.parse(event.data);
+            setEvents((_events) => [ data, ..._events ]);
```
[the remark this answers is not in the document]
Comment (irdkwmnsb) [reply admin/src/components/AnalyticsEvent.js:106]: Or perhaps not this, but the rerendering part. 
I would suggest not storing events in a state, but rather in a ref, and then after the debounce timer hits update the state with the ref value

**Review by irdkwmnsb — APPROVED** [review]
Minor fixed needed, but we should be fine.

File: admin/src/services/presetWidget.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { createApiGet, createApiPost } from "../utils";
+import { BASE_URL_BACKEND } from "../config";
+
+export class PresetWidgetService {
+    constructor(apiPath, errorHandler) {
+        this.apiUrl = BASE_URL_BACKEND + apiPath;
```
[the remark this answers is not in the document]
Comment (kbats183) [reply admin/src/services/presetWidget.js:6]: Yes, but it's different for different widgets. It may be "/advertisement" or "/title" or ...

**Pull request merged**